### PR TITLE
Container locking - restrict container to currently assigned domains

### PR DESCRIPTION
**Pull request by mckenfra**
#847 Provides functionality for locking a container. Once locked, any requests for domains that are not in the assignments list for that container will open in a new tab.

Unit tests included.

Container Unlocked

Container Locked

(This pull request is identical to the earlier one I submitted, but that one got cancelled because I moved the checkins to a separate branch on my fork.)

### CODE REVIEW

**Comment by jonathanKingston** [comment]
@maxxcrawford also before we merge it's probably worth confirming with product or Tanvi we are good to merge as it's a fair amount of functional change.

**Comment by mckenfra** [comment]
> Can you confirm that the icons (SVGs) in this PR are open-source and match the license used in this project?

The 2 new icons I added in my commit are taken (unchanged) from the MaterialUI project, github links here:

https://github.com/mui-org/material-ui/blob/9259aaaebbeded8b9b538a37c4be6094baede348/packages/material-ui-icons/src/Lock.js
https://github.com/mui-org/material-ui/blob/9259aaaebbeded8b9b538a37c4be6094baede348/packages/material-ui-icons/src/LockOpen.js

The licence used by the MaterialUI project is the MIT licence, github link here:

https://github.com/mui-org/material-ui/blob/9259aaaebbeded8b9b538a37c4be6094baede348/LICENSE

I believe the MIT licence is more permissive than Mozilla Public Licence 2.0 used by `mozilla/multi-account-containers`. It's certainly shorter! But the only issue I can see is that the MIT licence states:

> The above copyright notice and this permission notice shall be included in all
> copies or substantial portions of the Software.

Perhaps I need to actually insert the MIT licence as a comment at the top of both of the new SVG icon files. What do you think?

**Comment by mckenfra** [comment]
@maxxcrawford @jonathanKingston I have updated the README to explain that the 2 newly-added icons are covered by the MIT licence, and have also added a copy of the MIT licence to the project directory. I believe the MIT licence is compatible with this project's licence.

**Comment by maxxcrawford** [comment]
@mckenfra We already have a License file (https://github.com/mozilla/multi-account-containers/blob/master/LICENSE.txt), uses Mozilla's MPL 2.0 

@TanviHacks Can you review this PR too? Wanted to make sure this feature works for the roadmap.

**Comment by groovecoder** [comment]
@mckenfra - can you try to resolve the conflicts here? We're considering merging this. Going to loop our UX folks into this too.

**Comment by mckenfra** [comment]
> @mckenfra We'd like to update this PR and potentially get it into a future M-A-C release. (We'll have to consult with our legal & UX teams first on the LICENSE issues and if this UX works with their plans for Containers UX in Firefox.)
> 
> Will you have time to make the requested changes and updates? Or would you like @kendallcorner to pick this up from you?

Thanks a lot for carefully reviewing my changes! Yup, I'm happy to go through your requested updates. Sure, makes sense about consulting with your other teams too.

File: src/js/background/backgroundLogic.js [hunk 1]
```diff
@@ -229,7 +245,9 @@ const backgroundLogic = {
         hasHiddenTabs: !!containerState.hiddenTabs.length,
         hasOpenTabs: !!openTabs.length,
         numberOfHiddenTabs: containerState.hiddenTabs.length,
-        numberOfOpenTabs: openTabs.length
+        numberOfOpenTabs: openTabs.length,
+        // https://github.com/mozilla/multi-account-containers/issues/847
```
[the remark this answers is not in the document]
Comment (mckenfra) [reply src/js/background/backgroundLogic.js]: Sure, makes sense

File: src/js/popup.js [hunk 2]
```diff
@@ -980,6 +985,38 @@ Logic.registerPanel(P_CONTAINER_EDIT, {
...
+      const lockElement = document.createElement("div");
+      lockElement.innerHTML = escaped`
+        <img class="icon" src="${lockOrUnlockIcon}">
+        <div title="${lockOrUnlockLabel}">
+          ${lockOrUnlockLabel}
+        </div>`;
+      lockElement.classList.add("container-info-tab-row", "clickable", "container-lockorunlock", lockOrUnlockClass);
+      tableElement.appendChild(lockElement);
+      
+      const that = this;
```
[the remark this answers is not in the document]
Comment (mckenfra) [reply src/js/popup.js]: Well spotted! Yes I agree, there's no need for a `that` constant here. I'll change it to use `this`.

File: src/js/background/assignManager.js [hunk 3]
```diff
@@ -141,8 +141,21 @@ const assignManager = {
...
     }
     const userContextId = this.getUserContextIdFromCookieStore(tab);
-    if (!siteSettings
-        || userContextId === siteSettings.userContextId
+    
+    // Determine if "locked out", i.e.:
+    // This request's URL is not associated with any particular contextualIdentity.
+    // But the current tab's contextualIdentity is locked. So must open request in new tab.
+    // https://github.com/mozilla/multi-account-containers/issues/847
+    let isLockedOut;
```
[the remark this answers is not in the document]
Comment (mckenfra) [reply src/js/background/assignManager.js]: Yes, I can rename all of the variable and function names from lock/unlock to something else. But my UI icons are very much lock/unlock symbols. Are you thinking of changing these too? Might it get a bit confusing if the icons are locks, but the code has references to "strict/relaxed"? But I do agree that getting the naming right in the code is important, so it's clear for future devs.

[the remark this answers is not in the document]
Comment (mckenfra) [reply src/js/background/assignManager.js]: Done: https://github.com/mozilla/multi-account-containers/pull/1474/commits/25f16c731795855aa019d50a018739d1b0b7f318

File: src/js/background/assignManager.js [hunk 4]
```diff
@@ -327,7 +347,20 @@ const assignManager = {
...
+      // Unlock container if no more assignments after this one is removed.
+      // https://github.com/mozilla/multi-account-containers/issues/847
+      const assignments = await this.storageArea.getByContainer(userContextId);
+      const assignmentKeys = Object.keys(assignments);
+      if (!(assignmentKeys.length > 1)) {
+        await backgroundLogic.lockOrUnlockContainer({
+          userContextId: userContextId,
+          isLocked: false
+        });
+      }
```
[the remark this answers is not in the document]
Comment (mckenfra) [reply src/js/background/assignManager.js]: Done: https://github.com/mozilla/multi-account-containers/pull/1474/commits/25f16c731795855aa019d50a018739d1b0b7f318

File: src/js/background/backgroundLogic.js [hunk 5]
```diff
@@ -123,6 +123,22 @@ const backgroundLogic = {
...
+    const cookieStoreId = this.cookieStoreId(options.userContextId);
+    const containerState = await identityState.storageArea.get(cookieStoreId);
+    if (options.isLocked) {
+      containerState.isLocked = "locked";
+    } else {
+      delete containerState.isLocked;
+    }
+    return await identityState.storageArea.set(cookieStoreId, containerState);
+  },
+
```
[the remark this answers is not in the document]
Comment (mckenfra) [reply src/js/background/backgroundLogic.js]: Shall we keep the locked/unlocked terminology? See improvements in: https://github.com/mozilla/multi-account-containers/pull/1474/commits/25f16c731795855aa019d50a018739d1b0b7f318

File: src/js/background/backgroundLogic.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (mckenfra) [reply src/js/background/backgroundLogic.js]: Done: https://github.com/mozilla/multi-account-containers/pull/1474/commits/25f16c731795855aa019d50a018739d1b0b7f318

File: src/js/background/messageHandler.js [hunk 6]
```diff
@@ -16,6 +16,10 @@ const messageHandler = {
       case "createOrUpdateContainer":
         response = backgroundLogic.createOrUpdateContainer(m.message);
         break;
+      case "lockOrUnlockContainer":
+        // https://github.com/mozilla/multi-account-containers/issues/847
```
[the remark this answers is not in the document]
Comment (mckenfra) [reply src/js/background/messageHandler.js]: Shall we keep the locked/unlocked terminology? See improvements in: https://github.com/mozilla/multi-account-containers/pull/1474/commits/25f16c731795855aa019d50a018739d1b0b7f318

File: src/js/popup.js [hunk 7]
```diff
@@ -4,6 +4,9 @@
 
 const CONTAINER_HIDE_SRC = "/img/container-hide.svg";
 const CONTAINER_UNHIDE_SRC = "/img/container-unhide.svg";
+// https://github.com/mozilla/multi-account-containers/issues/847
```
[the remark this answers is not in the document]
Comment (mckenfra) [reply src/js/popup.js]: Done: https://github.com/mozilla/multi-account-containers/pull/1474/commits/25f16c731795855aa019d50a018739d1b0b7f318

File: src/js/popup.js [hunk 8]
```diff
@@ -252,6 +255,8 @@ const Logic = {
         identity.hasHiddenTabs = stateObject.hasHiddenTabs;
         identity.numberOfHiddenTabs = stateObject.numberOfHiddenTabs;
         identity.numberOfOpenTabs = stateObject.numberOfOpenTabs;
+        // https://github.com/mozilla/multi-account-containers/issues/847
```
[the remark this answers is not in the document]
Comment (mckenfra) [reply src/js/popup.js]: Done: https://github.com/mozilla/multi-account-containers/pull/1474/commits/25f16c731795855aa019d50a018739d1b0b7f318

File: src/js/popup.js [hunk 9]
```diff
@@ -969,7 +974,7 @@ Logic.registerPanel(P_CONTAINER_EDIT, {
     }
   },
 
-  showAssignedContainers(assignments) {
+  showAssignedContainers(assignments, isLocked) {
```
[the remark this answers is not in the document]
Comment (mckenfra) [reply src/js/popup.js]: Shall we keep the locked/unlocked terminology? See improvements in: https://github.com/mozilla/multi-account-containers/pull/1474/commits/25f16c731795855aa019d50a018739d1b0b7f318

File: src/js/popup.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (mckenfra) [reply src/js/popup.js]: Done: https://github.com/mozilla/multi-account-containers/pull/1474/commits/25f16c731795855aa019d50a018739d1b0b7f318

**Comment by mckenfra** [comment]
> @mckenfra - can you try to resolve the conflicts here? We're considering merging this. Going to loop our UX folks into this too.

I'll resolve the conflicts as the last step, once you're happy with the other changes. Does that sound good?

File: src/js/popup.js [hunk 10]
```diff
@@ -252,6 +255,8 @@ const Logic = {
         identity.hasHiddenTabs = stateObject.hasHiddenTabs;
         identity.numberOfHiddenTabs = stateObject.numberOfHiddenTabs;
         identity.numberOfOpenTabs = stateObject.numberOfOpenTabs;
+        // https://github.com/mozilla/multi-account-containers/issues/847
+        identity.isLocked = stateObject.isLocked;
```
[the remark this answers is not in the document]
Comment (mckenfra) [reply src/js/popup.js]: Shall we keep the locked/unlocked terminology? See improvements in: https://github.com/mozilla/multi-account-containers/pull/1474/commits/25f16c731795855aa019d50a018739d1b0b7f318

**Comment by mckenfra** [comment]
> > @mckenfra - can you try to resolve the conflicts here? We're considering merging this. Going to loop our UX folks into this too.
> 
> I'll resolve the conflicts as the last step, once you're happy with the other changes. Does that sound good?

Done

**Comment by kendallcorner** [comment]
Hi @mckenfra. This code has some merge conflicts with master right now, and I've even been working on some UX updates that (hopefully) should get merged soon that will create even more conflicts. *But* I *really* wanted to merge in this PR, so what I would like to do is copy the relevant code from this PR into my UX update PR and add you as a co-author. Would that be ok?

**Comment by mckenfra** [comment]
> Hi @mckenfra. This code has some merge conflicts with master right now, and I've even been working on some UX updates that (hopefully) should get merged soon that will create even more conflicts. _But_ I _really_ wanted to merge in this PR, so what I would like to do is copy the relevant code from this PR into my UX update PR and add you as a co-author. Would that be ok?

@kendallcorner  Sure, go ahead!

**Comment by kendallcorner** [comment]
Hi @mckenfra. This code has some merge conflicts with master right now, and I've even been working on some UX updates that (hopefully) should get merged soon that will create even more conflicts. *But* I *really* wanted to merge in this PR, so what I would like to do is copy the relevant code from this PR into my UX update PR and add you as a co-author. Would that be ok?

**Comment by rubaboo** [comment]
So how's it going, guys? MAC will make much more sense when this functionality is added. Thank you.

**Comment by JonRiv** [comment]
> So how's it going, guys? MAC will make much more sense when this functionality is added. Thank you.

+1 this is a most wanted feature ; without it behavior is a bit crippled if you actually want to protect cookies from your current container.

**Comment by joweiser** [comment]
Only one last nit pick to fix over here: https://github.com/mozilla/multi-account-containers/pull/1654
Looks like this will be merged very soon - thanks @mckenfra and @kendallcorner for your work on this!

**Pull request closed without merging**